File: src/main/java/Solution_350_IntersectionOfTwoArraysII.java
import java.util.ArrayList;
import java.util.Arrays;

public class Solution_350_IntersectionOfTwoArraysII {
    public static void main(String[] args) {
        var nums1 = new int[]{0, 1, 1, 2, 2, 4};
        var nums2 = new int[]{1, 1, 3};

        intersect(nums1, nums2);
    }

    /**
     * Используем метод двух указателей.
     * Сортируем оба массива в порядке возрастания.
     * Выбираем меньший и больший по размеру массив.
     * Проходим по каждому числу из меньшего массива и ищем такое же или большее число в большем массиве.
     * Если число совпало, добавляем в результирующий список и двигаем указатель на 1 ячейку вправо.
     * Если j вышел за границы большего массива, значит поиск окончен
     */
    private static int[] intersect(int[] nums1, int[] nums2) {
        var result = new ArrayList<Integer>();
        // Сортируем оба массива в порядке возрастания
        Arrays.sort(nums1);
        Arrays.sort(nums2);

        // Выбираем меньший и больший по размеру массив
        var minArray = new int[]{};
        var maxArray = new int[]{};

        if (nums1.length < nums2.length) {
            minArray = nums1;
            maxArray = nums2;
        } else {
            minArray = nums2;
            maxArray = nums1;
        }


        // Проходим по каждому числу из меньшего массива
        for (int i = 0, j = 0; i < minArray.length; i++) {
            var v1 = minArray[i];
            // ищем такое же или большее число в большем массиве
            while (j < maxArray.length - 1 && v1 > maxArray[j])
                j++;

            // Если число совпало, добавляем в результирующий список и двигаем указатель на 1 ячейку вправо
            if (maxArray[j] == v1) {
                result.add(v1);
                j++;
            }

            // Если j вышел за границы большего массива, значит поиск окончен
            if (j >= maxArray.length) break;
        }

        return result.stream().mapToInt(Integer::intValue).toArray();
    }
}

File: src/main/kotlin/Solution_350_IntersectionOfTwoArraysII.kt
/**
 * 350 Intersection of Two Arrays II
 *
 * Easy
 *
 * Given two integer arrays nums1 and nums2, return an array of their intersection. Each element in the result must appear as many times as it shows in both arrays and you may return the result in any order.
 *
 *
 *
 * Example 1:
 *
 * Input: nums1 = [1,2,2,1], nums2 = [2,2]
 * Output: [2,2]
 * Example 2:
 *
 * Input: nums1 = [4,9,5], nums2 = [9,4,9,8,4]
 * Output: [4,9]
 * Explanation: [9,4] is also accepted.
 *
 *
 * Constraints:
 *
 * 1 <= nums1.length, nums2.length <= 1000
 * 0 <= nums1[i], nums2[i] <= 1000
 *
 *
 * Follow up:
 *
 * What if the given array is already sorted? How would you optimize your algorithm?
 * What if nums1's size is small compared to nums2's size? Which algorithm is better?
 * What if elements of nums2 are stored on disk, and the memory is limited such that you cannot load all elements into the memory at once?
 */
fun main() {
    val nums1 = intArrayOf(1, 2, 2, 1)
    val nums2 = intArrayOf(2, 2)
    intersect(nums1, nums2)
}

/**
 * Используем hashmap для хранения частот чисел в одном из массивов
 * Проходим по второму массиву и ищем в мапе такое же число.
 * Если число есть и его количество больше или равно 1,добавляем число в result и уменьшаем его количество на 1
 */
fun intersect(nums1: IntArray, nums2: IntArray): IntArray {
    // Используем hashmap для хранения частот чисел в одном из массивов
    val map = mutableMapOf<Int, Int>()
    val result = mutableListOf<Int>()

    // Считаем количество повторений каждого числа в первом массиве
    for (num in nums1) {
        map.merge(num, 1, Int::plus)
    }

    // Проходим по каждому числу второго массива и ищем его в мапе
    for (num in nums2) {
        val count = map[num]
        // если нашли, проверяем количество
        if (count != null && count >= 1) {
            // добавляем в ответ
            result.add(num)
            // уменьшаем счетчик числа
            map.merge(num, 1, Int::minus)
        }
    }

    return result.toIntArray()
}